File: app/modules/GHCPAgentHubCore/Sources/GHCPAgentHub/Services/SessionEventsParser.swift
//
//  SessionEventsParser.swift
//  GHCPAgentHub
//
//  Parser for GitHub Copilot CLI events.jsonl files
//  Adapted from AgentHub's SessionJSONLParser for Claude Code
//

import Foundation

// MARK: - SessionEventsParser

/// Parser for Copilot CLI session events.jsonl files that extracts monitoring data
public struct SessionEventsParser {

  // MARK: - Event Types (Copilot-specific)

  /// Raw event from Copilot events.jsonl file
  public struct SessionEvent: Decodable {
    let type: String
    let data: AnyCodable?
    let id: String?
    let timestamp: String?
    let parentId: String?
  }

  // MARK: - Parsing Results

  /// Result of parsing a session file
  public struct ParseResult {
    public var model: String?
    public var lastInputTokens: Int = 0
    public var lastOutputTokens: Int = 0
    public var totalOutputTokens: Int = 0
    public var cacheReadTokens: Int = 0
    public var cacheCreationTokens: Int = 0
    public var messageCount: Int = 0
    public var toolCalls: [String: Int] = [:]
    public var pendingToolUses: [String: PendingToolInfo] = [:]  // toolCallId -> info
    public var recentActivities: [ActivityEntry] = []
    public var lastActivityAt: Date?
    public var sessionStartedAt: Date?
    public var currentStatus: SessionStatus = .idle
    public var gitBranch: String?
    public var summary: String?
    public var cwd: String?
    public var copilotVersion: String?

    public init() {}
  }

  /// Info about a pending tool use
  public struct PendingToolInfo {
    public let toolName: String
    public let toolCallId: String
    public let timestamp: Date
    public let input: String?
    public let codeChangeInput: CodeChangeInput?
  }

  // MARK: - Public API

  /// Parse an entire session events file and return aggregated state
  /// - Parameters:
  ///   - path: Path to the session events.jsonl file
  ///   - approvalTimeoutSeconds: Seconds to wait before considering a tool as awaiting approval (default: 5)
  public static func parseSessionFile(at path: String, approvalTimeoutSeconds: Int = 5) -> ParseResult {
    var result = ParseResult()

    guard let data = FileManager.default.contents(atPath: path),
          let content = String(data: data, encoding: .utf8) else {
      return result
    }

    let lines = content.components(separatedBy: .newlines)

    for line in lines where !line.isEmpty {
      if let event = parseEvent(line) {
        processEvent(event, into: &result)
      }
    }

    // Determine current status from pending tools
    updateCurrentStatus(&result, approvalTimeoutSeconds: approvalTimeoutSeconds)

    return result
  }

  /// Parse new lines from a session file (for incremental updates)
  public static func parseNewLines(_ lines: [String], into result: inout ParseResult, approvalTimeoutSeconds: Int = 5) {
    for line in lines where !line.isEmpty {
      if let event = parseEvent(line) {
        processEvent(event, into: &result)
      }
    }
    updateCurrentStatus(&result, approvalTimeoutSeconds: approvalTimeoutSeconds)
  }

  /// Parse a single JSONL line
  public static func parseEvent(_ line: String) -> SessionEvent? {
    guard let data = line.data(using: .utf8) else { return nil }

    do {
      let decoder = JSONDecoder()
      return try decoder.decode(SessionEvent.self, from: data)
    } catch {
      // Many lines may not match our expected format
      return nil
    }
  }

  // MARK: - Private Processing

  private static func processEvent(_ event: SessionEvent, into result: inout ParseResult) {
    let timestamp = parseTimestamp(event.timestamp)

    // Track first/last activity
    if let ts = timestamp {
      if result.sessionStartedAt == nil {
        result.sessionStartedAt = ts
      }
      result.lastActivityAt = ts
    }

    // Extract data dictionary for most event types
    let eventData = event.data?.value as? [String: Any]

    // Process based on Copilot event type
    switch event.type {

    // Session lifecycle
    case "session.start":
      if let context = eventData?["context"] as? [String: Any] {
        result.gitBranch = context["branch"] as? String
        result.cwd = context["cwd"] as? String
      }
      result.copilotVersion = eventData?["copilotVersion"] as? String

    // User message
    case "user.message":
      result.messageCount += 1
      let content = eventData?["content"] as? String ?? ""
      if !content.isEmpty {
        addActivity(
          type: .userMessage,
          description: String(content.prefix(80)),
          timestamp: timestamp,
          to: &result
        )
      }

    // Assistant turn start
    case "assistant.turn_start":
      addActivity(
        type: .thinking,
        description: "Thinking...",
        timestamp: timestamp,
        to: &result
      )

    // Assistant message (may contain tool requests)
    case "assistant.message":
      result.messageCount += 1

      // Extract model if present
      if let model = eventData?["model"] as? String {
        result.model = model
      }

      // Process tool requests
      if let toolRequests = eventData?["toolRequests"] as? [[String: Any]] {
        for toolRequest in toolRequests {
          guard let name = toolRequest["name"] as? String,
                let toolCallId = toolRequest["toolCallId"] as? String else { continue }

          // Track tool call count
          result.toolCalls[name, default: 0] += 1

          // Extract input preview
          let arguments = toolRequest["arguments"] as? [String: Any]
          let inputPreview = extractInputPreview(arguments)
          let codeChangeInput = extractCodeChangeInput(name: name, input: arguments)

          // Add to pending
          result.pendingToolUses[toolCallId] = PendingToolInfo(
            toolName: name,
            toolCallId: toolCallId,
            timestamp: timestamp ?? Date(),
            input: inputPreview,
            codeChangeInput: codeChangeInput
          )

          addActivity(
            type: .toolUse(name: name),
            description: inputPreview ?? name,
            timestamp: timestamp,
            codeChangeInput: codeChangeInput,
            to: &result
          )
        }
      }

      // Check for text content (assistant response)
      if let content = eventData?["content"] as? String, !content.isEmpty {
        addActivity(
          type: .assistantMessage,
          description: String(content.prefix(50)),
          timestamp: timestamp,
          to: &result
        )
      }

    // Tool result
    case "tool.result":
      if let toolCallId = eventData?["toolCallId"] as? String {
        let toolName = result.pendingToolUses[toolCallId]?.toolName ?? "unknown"
        result.pendingToolUses.removeValue(forKey: toolCallId)

        let isError = eventData?["isError"] as? Bool ?? false

        addActivity(
          type: .toolResult(name: toolName, success: !isError),
          description: isError ? "Error" : "Completed",
          timestamp: timestamp,
          to: &result
        )
      }

    // Assistant turn end
    case "assistant.turn_end":
      // Turn completed - may include usage info
      if let usage = eventData?["usage"] as? [String: Any] {
        result.lastInputTokens = usage["inputTokens"] as? Int ?? 0
        result.lastOutputTokens = usage["outputTokens"] as? Int ?? 0
        result.totalOutputTokens += result.lastOutputTokens
        result.cacheReadTokens += usage["cacheReadInputTokens"] as? Int ?? 0
        result.cacheCreationTokens += usage["cacheCreationInputTokens"] as? Int ?? 0
      }

    // Session info (e.g., MCP server info)
    case "session.info":
      // Informational, could log if needed
      break

    default:
      break
    }
  }

  /// Re-evaluate current status based on time elapsed since last activity
  public static func updateCurrentStatus(_ result: inout ParseResult, approvalTimeoutSeconds: Int = 5) {
    guard let lastActivity = result.recentActivities.last else {
      result.currentStatus = .idle
      return
    }

    let timeSince = Date().timeIntervalSince(lastActivity.timestamp)

    // Global idle timeout: 5 minutes
    if timeSince > 300 {
      result.currentStatus = .idle
      return
    }

    switch lastActivity.type {
    case .toolUse(let name):
      // Task tool runs in background, doesn't need approval
      if name == "task" || name == "Task" {
        result.currentStatus = .executingTool(name: name)
      } else if timeSince > Double(approvalTimeoutSeconds) {
        result.currentStatus = .awaitingApproval(tool: name)
      } else {
        result.currentStatus = .executingTool(name: name)
      }

    case .toolResult:
      if timeSince < 60 {
        result.currentStatus = .thinking
      } else {
        result.currentStatus = .idle
      }

    case .assistantMessage:
      result.currentStatus = .waitingForUser

    case .userMessage:
      if timeSince < 60 {
        result.currentStatus = .thinking
      } else {
        result.currentStatus = .idle
      }

    case .thinking:
      if timeSince < 30 {
        result.currentStatus = .thinking
      } else {
        result.currentStatus = .idle
      }
    }
  }

  private static func addActivity(
    type: ActivityType,
    description: String,
    timestamp: Date?,
    codeChangeInput: CodeChangeInput? = nil,
    to result: inout ParseResult
  ) {
    let entry = ActivityEntry(
      timestamp: timestamp ?? Date(),
      type: type,
      description: description,
      toolInput: codeChangeInput
    )

    result.recentActivities.append(entry)

    // Keep more activities for tracking (100)
    if result.recentActivities.count > 100 {
      result.recentActivities.removeFirst(result.recentActivities.count - 100)
    }
  }

  /// Extract full input parameters for code-changing tools
  private static func extractCodeChangeInput(name: String, input: [String: Any]?) -> CodeChangeInput? {
    guard let input = input,
          let filePath = input["file_path"] as? String ?? input["path"] as? String else {
      return nil
    }

    switch name {
    case "edit", "Edit":
      return CodeChangeInput(
        toolType: .edit,
        filePath: filePath,
        oldString: input["old_string"] as? String ?? input["old_str"] as? String,
        newString: input["new_string"] as? String ?? input["new_str"] as? String,
        replaceAll: input["replace_all"] as? Bool
      )

    case "write", "Write", "create", "Create":
      return CodeChangeInput(
        toolType: .write,
        filePath: filePath,
        newString: input["content"] as? String ?? input["file_text"] as? String
      )

    case "MultiEdit", "multi_edit":
      var editsArray: [[String: String]]? = nil
      if let edits = input["edits"] as? [[String: Any]] {
        editsArray = edits.compactMap { edit in
          var result = [String: String]()
          if let oldStr = edit["old_string"] as? String { result["old_string"] = oldStr }
          if let newStr = edit["new_string"] as? String { result["new_string"] = newStr }
          if let replaceAll = edit["replace_all"] as? Bool { result["replace_all"] = String(replaceAll) }
          return result.isEmpty ? nil : result
        }
      }
      return CodeChangeInput(
        toolType: .multiEdit,
        filePath: filePath,
        edits: editsArray
      )

    default:
      return nil
    }
  }

  // MARK: - Helpers

  private static func parseTimestamp(_ string: String?) -> Date? {
    guard let string = string else { return nil }

    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]

    if let date = formatter.date(from: string) {
      return date
    }

    // Try without fractional seconds
    formatter.formatOptions = [.withInternetDateTime]
    return formatter.date(from: string)
  }

  private static func extractInputPreview(_ input: [String: Any]?) -> String? {
    guard let input = input else { return nil }

    // Common patterns
    if let path = input["file_path"] as? String ?? input["path"] as? String {
      return URL(fileURLWithPath: path).lastPathComponent
    }
    if let command = input["command"] as? String {
      return String(command.prefix(50))
    }
    if let pattern = input["pattern"] as? String {
      return pattern
    }
    if let query = input["query"] as? String {
      return String(query.prefix(50))
    }
    if let intent = input["intent"] as? String {
      return String(intent.prefix(50))
    }

    return nil
  }
}

// MARK: - AnyCodable

/// Type-erased Codable wrapper for arbitrary JSON values
public struct AnyCodable: Decodable {
  public let value: Any

  public init(from decoder: Decoder) throws {
    let container = try decoder.singleValueContainer()

    if container.decodeNil() {
      self.value = NSNull()
    } else if let bool = try? container.decode(Bool.self) {
      self.value = bool
    } else if let int = try? container.decode(Int.self) {
      self.value = int
    } else if let double = try? container.decode(Double.self) {
      self.value = double
    } else if let string = try? container.decode(String.self) {
      self.value = string
    } else if let array = try? container.decode([AnyCodable].self) {
      self.value = array.map { $0.value }
    } else if let dictionary = try? container.decode([String: AnyCodable].self) {
      self.value = dictionary.mapValues { $0.value }
    } else {
      throw DecodingError.dataCorruptedError(
        in: container,
        debugDescription: "Unable to decode value"
      )
    }
  }
}

File: electron-app/src/services/SessionEventsParser.ts
/**
 * SessionEventsParser
 * 
 * Parser for GitHub Copilot CLI events.jsonl files.
 * Port of the Swift SessionEventsParser for parity with macOS version.
 */

import * as fs from 'fs';
import * as path from 'path';
import {
  SessionStatus,
  ActivityEntry,
  ActivityType,
  CodeChangeInput,
  PendingToolUse,
  PendingUserQuestion,
} from '../models/types';

/** Per-file parse cache keyed by absolute path */
interface ParseCache {
  result: ParseResult;
  mtime: number;
  size: number;
}
const _parseCache = new Map<string, ParseCache>();

/**
 * Raw event from Copilot events.jsonl
 */
interface SessionEvent {
  type: string;
  data?: Record<string, unknown>;
  id?: string;
  timestamp?: string;
  parentId?: string | null;
}

/**
 * Pending tool info (internal)
 */
interface PendingToolInfo {
  toolName: string;
  toolCallId: string;
  timestamp: Date;
  input?: string;
  codeChangeInput?: CodeChangeInput;
}

/**
 * Parse result - aggregated session state
 */
export interface ParseResult {
  model?: string;
  lastInputTokens: number;
  lastOutputTokens: number;
  totalOutputTokens: number;
  cacheReadTokens: number;
  cacheCreationTokens: number;
  messageCount: number;
  toolCalls: Record<string, number>;
  pendingToolUses: Map<string, PendingToolInfo>;
  pendingQuestion?: PendingUserQuestion;
  recentActivities: ActivityEntry[];
  lastActivityAt?: Date;
  sessionStartedAt?: Date;
  currentStatus: SessionStatus;
  gitBranch?: string;
  summary?: string;
  cwd?: string;
  copilotVersion?: string;
}

/**
 * Create empty parse result
 */
function createEmptyResult(): ParseResult {
  return {
    lastInputTokens: 0,
    lastOutputTokens: 0,
    totalOutputTokens: 0,
    cacheReadTokens: 0,
    cacheCreationTokens: 0,
    messageCount: 0,
    toolCalls: {},
    pendingToolUses: new Map(),
    recentActivities: [],
    currentStatus: { type: 'idle' },
  };
}

/**
 * Parse ISO8601 timestamp
 */
function parseTimestamp(ts?: string): Date | undefined {
  if (!ts) return undefined;
  try {
    return new Date(ts);
  } catch {
    return undefined;
  }
}

/**
 * Parse a single event line
 */
export function parseEvent(line: string): SessionEvent | null {
  try {
    return JSON.parse(line) as SessionEvent;
  } catch {
    return null;
  }
}

/**
 * Extract input preview from tool arguments
 */
function extractInputPreview(args?: Record<string, unknown>): string | undefined {
  if (!args) return undefined;

  // Common patterns (mirrors Swift implementation)
  if (typeof args.file_path === 'string') {
    return path.basename(args.file_path);
  }
  if (typeof args.path === 'string') {
    return path.basename(args.path);
  }
  if (typeof args.command === 'string') {
    return (args.command as string).substring(0, 50);
  }
  if (typeof args.pattern === 'string') {
    return args.pattern as string;
  }
  if (typeof args.query === 'string') {
    return (args.query as string).substring(0, 50);
  }
  if (typeof args.intent === 'string') {
    return (args.intent as string).substring(0, 50);
  }

  return undefined;
}

/**
 * Extract code change input for edit/write tools
 */
function extractCodeChangeInput(
  name: string,
  args?: Record<string, unknown>
): CodeChangeInput | undefined {
  if (!args) return undefined;

  const filePath = (args.file_path ?? args.path) as string | undefined;
  if (!filePath) return undefined;

  const nameLower = name.toLowerCase();

  if (nameLower === 'edit') {
    return {
      toolType: 'edit',
      filePath,
      oldString: (args.old_string ?? args.old_str) as string | undefined,
      newString: (args.new_string ?? args.new_str) as string | undefined,
      replaceAll: args.replace_all as boolean | undefined,
    };
  }

  if (nameLower === 'write' || nameLower === 'create') {
    return {
      toolType: 'write',
      filePath,
      newString: (args.content ?? args.file_text) as string | undefined,
    };
  }

  if (nameLower === 'multiedit' || nameLower === 'multi_edit') {
    const edits = args.edits as Array<Record<string, unknown>> | undefined;
    return {
      toolType: 'multiEdit',
      filePath,
      edits: edits?.map(e => ({
        old_string: e.old_string as string | undefined,
        new_string: e.new_string as string | undefined,
        replace_all: e.replace_all?.toString(),
      })),
    };
  }

  return undefined;
}

/**
 * Add activity to result (keeps last 100)
 */
function addActivity(
  result: ParseResult,
  activityType: ActivityType,
  description: string,
  timestamp?: Date,
  codeChangeInput?: CodeChangeInput
): void {
  result.recentActivities.push({
    timestamp: timestamp ?? new Date(),
    activityType,
    description,
    toolInput: codeChangeInput,
  });

  // Keep only last 100 activities (mirrors Swift implementation)
  if (result.recentActivities.length > 100) {
    result.recentActivities = result.recentActivities.slice(-100);
  }
}

/**
 * Process a single event into the result
 */
function processEvent(event: SessionEvent, result: ParseResult): void {
  const timestamp = parseTimestamp(event.timestamp);
  const data = event.data ?? {};

  // Track timestamps
  if (timestamp) {
    if (!result.sessionStartedAt) {
      result.sessionStartedAt = timestamp;
    }
    result.lastActivityAt = timestamp;
  }

  // Process by event type (mirrors Swift implementation)
  switch (event.type) {
    case 'session.start': {
      const context = data.context as Record<string, unknown> | undefined;
      if (context) {
        result.gitBranch = context.branch as string | undefined;
        result.cwd = context.cwd as string | undefined;
      }
      result.copilotVersion = data.copilotVersion as string | undefined;
      break;
    }

    case 'user.message': {
      result.messageCount++;
      const content = (data.content as string) ?? '';
      if (content) {
        addActivity(result, { type: 'userMessage' }, content.substring(0, 80), timestamp);
      }
      break;
    }

    case 'assistant.turn_start': {
      addActivity(result, { type: 'thinking' }, 'Thinking...', timestamp);
      break;
    }

    case 'assistant.message': {
      result.messageCount++;

      // Extract model
      if (data.model) {
        result.model = data.model as string;
      }

      // Process tool requests
      const toolRequests = data.toolRequests as Array<Record<string, unknown>> | undefined;
      if (toolRequests) {
        for (const toolReq of toolRequests) {
          const name = toolReq.name as string | undefined;
          const toolCallId = toolReq.toolCallId as string | undefined;

          if (name && toolCallId) {
            // Track tool call count
            result.toolCalls[name] = (result.toolCalls[name] ?? 0) + 1;

            // Extract input
            const args = toolReq.arguments as Record<string, unknown> | undefined;
            const inputPreview = extractInputPreview(args);
            const codeChangeInput = extractCodeChangeInput(name, args);

            // Check for ask_user tool - track pending question
            if (name === 'ask_user' && args) {
              result.pendingQuestion = {
                toolCallId,
                question: (args.question as string) ?? 'Question',
                choices: args.choices as string[] | undefined,
                timestamp: timestamp ?? new Date(),
              };
            }

            // Add to pending
            result.pendingToolUses.set(toolCallId, {
              toolName: name,
              toolCallId,
              timestamp: timestamp ?? new Date(),
              input: inputPreview,
              codeChangeInput,
            });

            addActivity(
              result,
              { type: 'toolUse', name },
              inputPreview ?? name,
              timestamp,
              codeChangeInput
            );
          }
        }
      }

      // Check for text content
      const content = data.content as string | undefined;
      if (content) {
        addActivity(result, { type: 'assistantMessage' }, content.substring(0, 50), timestamp);
      }
      break;
    }

    case 'tool.result': {
      const toolCallId = data.toolCallId as string | undefined;
      if (toolCallId) {
        const pending = result.pendingToolUses.get(toolCallId);
        const toolName = pending?.toolName ?? 'unknown';
        result.pendingToolUses.delete(toolCallId);

        // Clear pending question if this was the ask_user response
        if (result.pendingQuestion?.toolCallId === toolCallId) {
          result.pendingQuestion = undefined;
        }

        const isError = data.isError as boolean | undefined ?? false;

        addActivity(
          result,
          { type: 'toolResult', name: toolName, success: !isError },
          isError ? 'Error' : 'Completed',
          timestamp
        );
      }
      break;
    }

    case 'assistant.turn_end': {
      // Extract usage info
      const usage = data.usage as Record<string, unknown> | undefined;
      if (usage) {
        result.lastInputTokens = (usage.inputTokens as number) ?? 0;
        result.lastOutputTokens = (usage.outputTokens as number) ?? 0;
        result.totalOutputTokens += result.lastOutputTokens;
        result.cacheReadTokens += (usage.cacheReadInputTokens as number) ?? 0;
        result.cacheCreationTokens += (usage.cacheCreationInputTokens as number) ?? 0;
      }
      break;
    }
  }
}

/**
 * Update current status based on recent activity
 * Mirrors Swift implementation for parity
 */
export function updateCurrentStatus(result: ParseResult, approvalTimeoutSeconds = 5): void {
  const lastActivity = result.recentActivities[result.recentActivities.length - 1];

  if (!lastActivity) {
    result.currentStatus = { type: 'idle' };
    return;
  }

  const timeSince = (Date.now() - lastActivity.timestamp.getTime()) / 1000;

  // Global idle timeout: 5 minutes (300 seconds)
  if (timeSince > 300) {
    result.currentStatus = { type: 'idle' };
    return;
  }

  switch (lastActivity.activityType.type) {
    case 'toolUse': {
      const name = lastActivity.activityType.name;
      // Task tool runs in background, doesn't need approval
      if (name.toLowerCase() === 'task') {
        result.currentStatus = { type: 'executingTool', name };
      } else if (timeSince > approvalTimeoutSeconds) {
        result.currentStatus = { type: 'awaitingApproval', tool: name };
      } else {
        result.currentStatus = { type: 'executingTool', name };
      }
      break;
    }

    case 'toolResult':
      result.currentStatus = timeSince < 60 ? { type: 'thinking' } : { type: 'idle' };
      break;

    case 'assistantMessage':
      result.currentStatus = { type: 'waitingForUser' };
      break;

    case 'userMessage':
      result.currentStatus = timeSince < 60 ? { type: 'thinking' } : { type: 'idle' };
      break;

    case 'thinking':
      result.currentStatus = timeSince < 30 ? { type: 'thinking' } : { type: 'idle' };
      break;
  }
}

/**
 * Parse an entire session file, with incremental caching by mtime+size.
 * Re-reads only new bytes when the file has grown.
 */
export function parseSessionFile(filePath: string, approvalTimeoutSeconds = 5): ParseResult {
  if (!fs.existsSync(filePath)) {
    return createEmptyResult();
  }

  let stat: fs.Stats;
  try {
    stat = fs.statSync(filePath);
  } catch {
    return createEmptyResult();
  }

  const mtime = stat.mtimeMs;
  const size = stat.size;
  const cached = _parseCache.get(filePath);

  // Fully cached — mtime and size unchanged
  if (cached && cached.mtime === mtime && cached.size === size) {
    return cached.result;
  }

  // Incremental: file grew since last parse — read only new bytes
  if (cached && size > cached.size && cached.mtime <= mtime) {
    let newContent = '';
    try {
      const fd = fs.openSync(filePath, 'r');
      const bytesToRead = size - cached.size;
      const buf = Buffer.allocUnsafe(bytesToRead);
      fs.readSync(fd, buf, 0, bytesToRead, cached.size);
      fs.closeSync(fd);
      newContent = buf.toString('utf-8');
    } catch {
      // Fall through to full re-parse on error
    }
    if (newContent) {
      const newLines = newContent.split('\n');
      parseNewLines(newLines, cached.result, approvalTimeoutSeconds);
      cached.mtime = mtime;
      cached.size = size;
      return cached.result;
    }
  }

  // Full parse
  const result = createEmptyResult();
  const content = fs.readFileSync(filePath, 'utf-8');
  const lines = content.split('\n').filter(line => line.trim());

  for (const line of lines) {
    const event = parseEvent(line);
    if (event) {
      processEvent(event, result);
    }
  }

  updateCurrentStatus(result, approvalTimeoutSeconds);
  _parseCache.set(filePath, { result, mtime, size });
  return result;
}

/**
 * Evict a session's parse cache entry (call when session is deleted).
 */
export function evictParseCache(filePath: string): void {
  _parseCache.delete(filePath);
}

/**
 * Parse new lines incrementally
 */
export function parseNewLines(
  lines: string[],
  result: ParseResult,
  approvalTimeoutSeconds = 5
): void {
  for (const line of lines) {
    if (!line.trim()) continue;
    const event = parseEvent(line);
    if (event) {
      processEvent(event, result);
    }
  }

  updateCurrentStatus(result, approvalTimeoutSeconds);
}
